src/model/analysis.ts

```diff
@@ -141,6 +144,30 @@ export class Analysis implements Queryable {
     return Array.from(result);
   }
 
+  /**
+   * Potentially narrow down the document that contains the sourceRange.
+   * For example, if a source range is inside a inlineDocument, this function
+   * will narrow down the document to the most specific inline document.
+   *
+   * @param sourceRange Source range to search for in a document
+   */
+   getDocumentContaining(sourceRange: SourceRange|undefined) {
+   if (!sourceRange) {
+     return undefined;
+   }
+   let mostSpecificDocument: undefined|Document = undefined;
+   for (const doc of this.getFeatures({kind: 'document'})) {
```

Comment:
(moving this comment from the other collapsed thread because github is bad at code review)

This goes over all documents, which could be hundreds or thousands. We can get the toplevel doc like so:

```javascript
const [outerDoc] = this.getFeatures({kind: 'document', id: sourceRange.file}));
if (outerDoc === undefined) {
  return undefined;
}
for (const doc of outerDoc.getFeatures({kind: 'document'})) {
   ...
}
```

Reply:
Ah yes that was exactly what I was looking for!